crates/c-api/src/wasi.rs

```diff
@@ -0,0 +1,274 @@
+//! The WASI embedding API definitions for Wasmtime.
+use crate::{wasm_extern_t, wasm_importtype_t, wasm_store_t, wasm_trap_t, ExternHost, ExternType};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::fs::File;
+use std::os::raw::{c_char, c_int};
+use std::path::Path;
+use std::slice;
+use wasi_common::{preopen_dir, WasiCtxBuilder};
+use wasmtime::{HostRef, Trap};
+use wasmtime_wasi::Wasi;
+
+unsafe fn cstr_to_path<'a>(path: *const c_char) -> Option<&'a Path> {
+    CStr::from_ptr(path).to_str().map(Path::new).ok()
+}
+
+unsafe fn open_file(path: *const c_char) -> Option<File> {
+    File::open(cstr_to_path(path)?).ok()
+}
+
+unsafe fn create_file(path: *const c_char) -> Option<File> {
+    File::create(cstr_to_path(path)?).ok()
+}
+
+#[repr(C)]
+pub struct wasi_config_t {
+    builder: WasiCtxBuilder,
+}
+
+impl wasi_config_t {}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_new() -> *mut wasi_config_t {
+    Box::into_raw(Box::new(wasi_config_t {
+        builder: WasiCtxBuilder::new(),
+    }))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_delete(config: *mut wasi_config_t) {
+    drop(Box::from_raw(config));
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_set_argv(
+    config: *mut wasi_config_t,
+    argc: c_int,
+    argv: *const *const c_char,
+) {
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.args(
+        slice::from_raw_parts(argv, argc as usize)
+            .iter()
+            .map(|a| slice::from_raw_parts(*a as *const u8, CStr::from_ptr(*a).to_bytes().len())),
+    );
+    std::mem::forget(config);
```

Comment:
This seems sort of odd because if a panic or something bad happens we may accidentally try to drop the config, could this work on `&mut wasi_config_t` instead?

(I'd be in favor of updating the builder to not consume `self`, but intead take and return `&mut self` to help with this situation)

Reply:
Isn't `WasiCtxBuilder` is a move-pattern builder for a reason though?  I really don't like the `forget` calls either and would be absolutely onboard with removing them.  I can look into making `WasiCtxBuilder` a mut-ref-pattern builder, but I think I took a stab at that originally and just decided to live with the unsoundness of `forget`.

Reply:
I think the builder may require by-val today, but we can fix that with internal `Option` values in the builder itself (or making the consumption method be by-value but builder methods by-reference)

Reply:
I've fixed up `WasiCtxBuilder` to take `&mut self`.  Please review in case there's a better way to do it.  Thanks!